File: src/micapolos/tata8/Key.java
package micapolos.tata8;

public final class Key {
  public boolean wasPressed;
  public boolean isPressed;

  public boolean isPressed() {
    return isPressed;
  }

  public boolean pressed() {
    return !wasPressed && isPressed;
  }

  public boolean released() {
    return wasPressed && !isPressed;
  }

  void press() {
    isPressed = true;
  }

  void release() {
    isPressed = false;
  }

  void update() {
    wasPressed = isPressed;
  }
}

File: src/micapolos/zexy2/Key.kt
package micapolos.zexy2

import micapolos.tata8.Game
import micapolos.zexy2.ast.Live
import micapolos.zexy2.ast.Primitive

enum class Key(internal val tata8: micapolos.tata8.Key) {
  LEFT(Game.keys.left),
  RIGHT(Game.keys.right),
  UP(Game.keys.up),
  DOWN(Game.keys.down),
  Z(Game.keys.z),
  X(Game.keys.x);

  val isPressed: Live<Boolean> =
    Live.Application(
      Boolean::class,
      Primitive.KEY_IS_PRESSED,
      listOf(constant(Key::class, this)))

  val pressed: Live<Boolean> =
    Live.Application(
      Boolean::class,
      Primitive.KEY_PRESSED,
      listOf(constant(Key::class, this)))

  val released: Live<Boolean> =
    Live.Application(
      Boolean::class,
      Primitive.KEY_RELEASED,
      listOf(constant(Key::class, this)))
}
